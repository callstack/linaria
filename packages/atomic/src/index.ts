export { default as css } from './css';
export { default as atomize } from './atomize';
export { cx } from '@linaria/utils';

export type { CSSProperties } from './CSSProperties';
export type { StyleCollectionObject } from './css';
