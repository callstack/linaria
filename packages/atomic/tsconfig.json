{
  "extends": "../../tsconfig.json",
  "compilerOptions": { "paths": {}, "rootDir": "src/" },
  "references": [{ "path": "../utils" }]
}
