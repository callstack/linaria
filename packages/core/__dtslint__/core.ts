/* eslint-disable @typescript-eslint/no-unused-vars */
import { css, cx } from '../src';

// $ExpectType LinariaClassName
const class1 = css``;

const activeClass = css``;

// $ExpectType string
const combined1 = cx(class1, 'active');

// $ExpectType LinariaClassName
const combined2 = cx(class1, activeClass);
