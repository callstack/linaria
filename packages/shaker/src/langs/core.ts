import { types as t } from '@babel/core';
import type {
  AssignmentExpression,
  Block,
  CallExpression,
  Directive,
  ExpressionStatement,
  ForInStatement,
  ForStatement,
  Function,
  Identifier,
  IfStatement,
  MemberExpression,
  Node,
  ObjectExpression,
  SequenceExpression,
  SwitchCase,
  SwitchStatement,
  Terminatorless,
  TryStatement,
  VariableDeclaration,
  VariableDeclarator,
  WhileStatement,
} from '@babel/types';

import { peek } from '@linaria/babel-preset';
import type { IdentifierHandlers, Visitors } from '../types';
import GraphBuilderState from '../GraphBuilderState';
import ScopeManager from '../scope';
import DepsGraph from '../DepsGraph';

function isIdentifier(
  node: Node,
  name?: string | string[]
): node is Identifier {
  return (
    t.isIdentifier(node) &&
    (name === undefined ||
      (Array.isArray(name) ? name.includes(node.name) : node.name === name))
  );
}

type SideEffect = [
  {
    callee?: (child: CallExpression['callee']) => boolean;
    arguments?: (child: CallExpression['arguments']) => boolean;
  },
  (node: CallExpression, state: GraphBuilderState) => void
];

const sideEffects: SideEffect[] = [
  [
    // if the first argument of forEach is required, mark forEach as required
    {
      callee: (node) =>
        t.isMemberExpression(node) &&
        t.isIdentifier(node.property) &&
        node.property.name === 'forEach',
    },
    (node, state) => state.graph.addEdge(node.arguments[0], node),
  ],
];

function getCallee(node: CallExpression): Node {
  if (
    t.isSequenceExpression(node.callee) &&
    node.callee.expressions.length === 2
  ) {
    const [first, second] = node.callee.expressions;
    if (t.isNumericLiteral(first) && first.value === 0) {
      return second;
    }
  }

  return node.callee;
}

function findWildcardReexportStatement(
  node: t.CallExpression,
  identifierName: string,
  graph: DepsGraph
): t.Statement | null {
  if (!t.isIdentifier(node.callee) || node.callee.name !== 'require')
    return null;

  const declarator = graph.getParent(node);
  if (!t.isVariableDeclarator(declarator)) return null;

  const declaration = graph.getParent(declarator);
  if (!t.isVariableDeclaration(declaration)) return null;

  const program = graph.getParent(declaration);
  if (!t.isProgram(program)) return null;

  // Our node is a correct export
  // Let's check that we have something that looks like transpiled re-export
  return (
    program.body.find((statement) => {
      /*
       * We are looking for `Object.keys(_bar).forEach(…)`
       */

      if (!t.isExpressionStatement(statement)) return false;

      const expression = statement.expression;
      if (!t.isCallExpression(expression)) return false;

      const callee = expression.callee;
      if (!t.isMemberExpression(callee)) return false;

      const { object, property } = callee;

      if (!isIdentifier(property, 'forEach')) return false;

      if (!t.isCallExpression(object)) return false;

      // `object` should be `Object.keys`
      if (
        !t.isMemberExpression(object.callee) ||
        !isIdentifier(object.callee.object, 'Object') ||
        !isIdentifier(object.callee.property, 'keys')
      )
        return false;

      //
      const [argument] = object.arguments;
      return isIdentifier(argument, identifierName);
    }) ?? null
  );
}

/*
 * Returns nodes which are implicitly affected by specified node
 */
function getAffectedNodes(node: Node, state: GraphBuilderState): Node[] {
  // FIXME: this method should be generalized
  const callee = t.isCallExpression(node) ? getCallee(node) : null;
  if (
    t.isCallExpression(node) &&
    t.isMemberExpression(callee) &&
    isIdentifier(callee.object, 'Object') &&
    isIdentifier(callee.property, [
      'assign',
      'defineProperty',
      'defineProperties',
      'freeze',
      'observe',
    ])
  ) {
    const [obj, property] = node.arguments;
    if (!t.isIdentifier(obj)) {
      return [];
    }

    if (
      state.scope.getDeclaration(obj) !== ScopeManager.globalExportsIdentifier
    ) {
      return [node.arguments[0]];
    }

    if (t.isStringLiteral(property)) {
      if (property.value === '__esModule') {
        return [node.arguments[0]];
      }

      state.graph.addExport(property.value, node);
    }
  }

  return [];
}

export const visitors: Visitors = {
  /*
   * FunctionDeclaration | FunctionExpression | ObjectMethod | ArrowFunctionExpression | ClassMethod | ClassPrivateMethod;
   * Functions can be either a statement or an expression.
   * That's why we need to disable default dependency resolving strategy for expressions by passing `ignoreDeps` flag.
   * Every function must have a body. Without a body, it becomes invalid.
   * In general, a body depends on parameters of a function.
   * In real life, some of the parameters can be omitted, but it's not trivial to implement that type of tree shaking.
   */
  Function(this: GraphBuilderState, node: Function) {
    const unsubscribe = this.onVisit((descendant) =>
      this.graph.addEdge(node, descendant)
    );
    this.baseVisit(node, true); // ignoreDeps=true prevents default dependency resolving
    unsubscribe();

    this.graph.addEdge(node, node.body);

    node.params.forEach((param) => this.graph.addEdge(node.body, param));
    if (
      t.isFunctionExpression(node) &&
      node.id !== null &&
      node.id !== undefined
    ) {
      // keep function name in expressions like `const a = function a();`
      this.graph.addEdge(node, node.id);
    }
  },

  /*
   * ExpressionStatement
   */
  ExpressionStatement(this: GraphBuilderState, node: ExpressionStatement) {
    this.baseVisit(node);

    this.graph.addEdge(node, node.expression);
    this.graph.addEdge(node.expression, node);
  },

  /*
   * BlockStatement | Program
   * The same situation as in ExpressionStatement: if one of the expressions is required, the block itself is also required.
   * Whereas a block doesn't depend on its children.
   * Example:
   * 1. let c;
   * 2. { // BlockStatement begin
   * 3.   let a = 1;
   * 4.   let b = 2;
   * 5.   a++;
   * 6.   a = c;
   * 7. } // BlockStatement end
   *
   * If we want to evaluate the value of `c`, we need to evaluate lines 1, 3, 5 and 6,
   * but we don't need line 4, even though it's a child of the block.
   */
  Block(this: GraphBuilderState, node: Block) {
    this.baseVisit(node);

    if (t.isProgram(node)) {
      const exportsDeclaration = this.scope.getDeclaration('global:exports')!;
      this.graph.addEdge(node, exportsDeclaration);
      node.directives.forEach((directive) =>
        this.graph.addEdge(node, directive)
      );
    }
  },

  Directive(this: GraphBuilderState, node: Directive) {
    this.baseVisit(node);
    this.graph.addEdge(node, node.value);
  },

  /*
   * TryStatement
   * try { /* block *\/ } catch() {/* handler *\/} finalize {/* finalizer *\/}
   * `handler` and `finalizer` do not make sense without `block`
   * `block` depends on the whole node.
   */
  TryStatement(this: GraphBuilderState, node: TryStatement) {
    this.baseVisit(node);
    [node.handler, node.finalizer].forEach((statement) => {
      if (statement) {
        this.graph.addEdge(node.block, statement);
        this.graph.addEdge(statement, node.block);
      }
    });
  },

  IfStatement(this: GraphBuilderState, node: IfStatement) {
    this.baseVisit(node);
    this.graph.addEdge(node, node.consequent);
    this.graph.addEdge(node, node.test);
  },

  /*
   * WhileStatement
   * Pretty simple behaviour here:
   * • if body is required, the statement is required
   * • if the statement is required, the condition is also required.
   */
  WhileStatement(this: GraphBuilderState, node: WhileStatement) {
    this.baseVisit(node);
    this.graph.addEdge(node, node.test);
  },

  SwitchCase(this: GraphBuilderState, node: SwitchCase) {
    this.baseVisit(node);
    node.consequent.forEach((statement) => this.graph.addEdge(statement, node));
    if (node.test) {
      this.graph.addEdge(node, node.test);
    }
  },

  SwitchStatement(this: GraphBuilderState, node: SwitchStatement) {
    this.baseVisit(node);
    node.cases.forEach((c) => this.graph.addEdge(c, node));
    this.graph.addEdge(node, node.discriminant);
  },

  ForStatement(this: GraphBuilderState, node: ForStatement) {
    this.baseVisit(node);

    [node.init, node.test, node.update, node.body].forEach((child) => {
      if (child) {
        this.graph.addEdge(node, child);
      }
    });
  },

  /*
   * ForInStatement
   * for (const k in o) { body }
   */
  ForInStatement(this: GraphBuilderState, node: ForInStatement) {
    this.baseVisit(node);

    if (node.body) {
      this.graph.addEdge(node, node.body);
      this.graph.addEdge(node.body, node.left);
    }

    this.graph.addEdge(node.left, node.right);
  },

  /*
   * BreakStatement | ContinueStatement | ReturnStatement | ThrowStatement | YieldExpression | AwaitExpression
   * All these nodes are required to evaluate the value of a function in which they are defined.
   * Also, the value of these nodes depends on the argument if it is presented.
   */
  Terminatorless(this: GraphBuilderState, node: Terminatorless) {
    this.baseVisit(node);

    if (
      !(t.isBreakStatement(node) || t.isContinueStatement(node)) &&
      node.argument
    ) {
      this.graph.addEdge(node, node.argument);
    }

    const closestFunctionNode = peek(this.fnStack);
    this.graph.addEdge(closestFunctionNode, node);
  },

  /*
   * ObjectExpression
   * Objects are… complicated. Especially because similarly looking code can be either an expression or a pattern.
   * In this case we work with an expression like:
   * const obj = {
   *   method() {}, // ObjectMethod
   *   property: "value", // ObjectProperty
   *   ...rest, // SpreadElement
   * }
   */
  ObjectExpression(this: GraphBuilderState, node: ObjectExpression) {
    this.context.push('expression');
    this.baseVisit(node);
    node.properties.forEach((prop) => {
      this.graph.addEdge(node, prop);
      if (t.isObjectMethod(prop)) {
        this.graph.addEdge(prop, prop.key);
        this.graph.addEdge(prop, prop.body);
      } else if (t.isObjectProperty(prop)) {
        this.graph.addEdge(prop, prop.key);
        this.graph.addEdge(prop, prop.value);
      } else if (t.isSpreadElement(prop)) {
        this.graph.addEdge(prop, prop.argument);
      }
    });
    this.context.pop();
  },

  /*
   * MemberExpression
   * It's about a simple expression like `obj.foo` or `obj['foo']`.
   * In addition to default behaviour (an expression depends on all its children),
   * we add a backward dependency from an object to a node for processing member
   * expressions in assignments.
   *
   * Example:
   * let obj = { a: 1 };
   * obj.b = 2;
   *
   * If we try to evaluate `obj` without backward dependency,
   * `obj.b = 2` will be cut and we will get just `{ a: 1 }`.
   */
  MemberExpression(this: GraphBuilderState, node: MemberExpression) {
    this.baseVisit(node);

    if (
      isIdentifier(node.object, 'exports') &&
      this.scope.getDeclaration(node.object) ===
        ScopeManager.globalExportsIdentifier
    ) {
      // We treat `exports.something` and `exports['something']` as identifiers in the global scope
      this.graph.addEdge(node, node.object);
      this.graph.addEdge(node, node.property);

      const isLVal = peek(this.context) === 'lval';
      if (isLVal) {
        this.scope.declare(node, false);
      } else {
        const declaration = this.scope.addReference(node);
        this.graph.addEdge(node, declaration);
      }

      return;
    }

    if (t.isIdentifier(node.object) && t.isIdentifier(node.property)) {
      // It's simple `foo.bar` expression. Is it a usage of a required library?
      const declaration = this.scope.getDeclaration(node.object);
      if (
        t.isIdentifier(declaration) &&
        this.graph.importAliases.has(declaration) &&
        !node.computed
      ) {
        // It is. We can remember what exactly we use from it.
        const source = this.graph.importAliases.get(declaration)!;
        this.graph.imports.get(source)!.push(node.property);
      }
    }
  },

  /*
   * AssignmentExpression
   * `a = b`, `{ ...rest } = obj`, `obj.a = 3`, etc.
   * It's not a declaration, it's just an assignment, but it affects
   * the value of declared variable if the variable it mentioned in the left part.
   * So, we apply some context-magic here in order to catch reference of variables in the left part.
   * We switch the context to `lval` and continue traversing through the left branch.
   * If we then meet some identifier, we mark it as a dependency of its declaration.
   */
  AssignmentExpression(this: GraphBuilderState, node: AssignmentExpression) {
    this.context.push('lval');
    this.visit<AssignmentExpression['left'], AssignmentExpression>(
      node.left,
      node,
      'left'
    );
    this.context.pop();

    this.visit(node.right, node, 'right');

    // The value of an expression depends on the left part.
    this.graph.addEdge(node, node.left);

    // The left part of an assignment depends on the right part.
    this.graph.addEdge(node.left, node.right);
  },

  /*
   * VariableDeclarator
   * It would be pretty simple if it weren't used to declare variables from other modules.
   */
  VariableDeclarator(this: GraphBuilderState, node: VariableDeclarator) {
    /*
     * declared is used for detecting external dependencies in cases like
     * const { a, b, c } = require('module');
     *
     * We are remembering all declared variables in order to use it later in CallExpression visitor
     */
    const declared: Array<[Identifier, Identifier | null]> = [];
    this.meta.set('declared', declared);
    const unregister = this.scope.addDeclareHandler((identifier, from) =>
      declared.push([identifier, from])
    );
    this.baseVisit(node);
    this.meta.delete('declared');
    unregister();

    if (node.init) {
      // If there is an initialization part, the identifier depends on it.
      this.graph.addEdge(node.id, node.init);
    }

    // If a statement is required itself, an id is also required
    this.graph.addEdge(node, node.id);
  },

  /*
   * VariableDeclaration
   * It's just a wrapper for group of VariableDeclarator.
   * If one of the declarators is required, the wrapper itself is also required.
   */
  VariableDeclaration(this: GraphBuilderState, node: VariableDeclaration) {
    this.meta.set('kind-of-declaration', node.kind);
    this.baseVisit(node);
    node.declarations.forEach((declaration) =>
      this.graph.addEdge(declaration, node)
    );
    this.meta.delete('kind-of-declaration');
  },

  /*
   * CallExpression
   * Do you remember that we have already mentioned it in VariableDeclarator?
   * It is a simple expression with default behaviour unless it is a `require`.
   *
   * Another tricky use case here is functions with side effects (e.g. `Object.defineProperty`).
   */
  CallExpression(
    this: GraphBuilderState,
    node: CallExpression,
    parent: Node | null
  ) {
    this.baseVisit(node);

    if (t.isIdentifier(node.callee) && node.callee.name === 'require') {
      // It looks like a module import …
      const scopeId = this.scope.whereIsDeclared(node.callee);
      if (scopeId && scopeId !== 'global') {
        // … but it is just a user defined function
        return;
      }

      const [firstArg] = node.arguments;
      if (!t.isStringLiteral(firstArg)) {
        // dynamic import? Maybe someday we can do something about it
        return;
      }

      const { value: source } = firstArg;
      const declared = this.meta.get('declared') as Array<
        [Identifier, Identifier | null]
      >;
      if (!declared) {
        // This is a standalone `require`
        return;
      }

      // Define all declared variables as external dependencies.
      declared.forEach(([local, _imported]) =>
        // FIXME: var slugify = require('../slugify').default;
        {
          if (!this.graph.imports.has(source)) {
            this.graph.imports.set(source, []);
          }

          if (
            parent &&
            t.isMemberExpression(parent) &&
            t.isIdentifier(parent.property)
          ) {
            // An imported function is specified right here.
            // eg. require('../slugify').default
            this.graph.imports.get(source)!.push(parent.property);
          } else {
            if (
              t.isCallExpression(parent) &&
              t.isIdentifier(parent.callee) &&
              typeof parent.callee.name === 'string'
            ) {
              if (parent.callee.name.startsWith('_interopRequireDefault')) {
                this.graph.importTypes.set(source, 'default');
              } else if (
                parent.callee.name.startsWith('_interopRequireWildcard')
              ) {
                this.graph.importTypes.set(source, 'wildcard');
              } else {
                // What I've missed?
              }
            }

            // Do we know the type of import?
            if (!this.graph.importTypes.has(source)) {
              // Is it a wildcard reexport? Let's check.
              const statement = findWildcardReexportStatement(
                node,
                local.name,
                this.graph
              );
              if (statement) {
                this.graph.addEdge(local, statement);
                this.graph.reexports.push(local);
                this.graph.importTypes.set(source, 'reexport');
              }
            }

            // The whole namespace was imported. We will know later, what exactly we need.
            // eg. const slugify = require('../slugify');
            this.graph.importAliases.set(local, source);
          }
        }
      );

      return;
    }

    sideEffects.forEach(([conditions, callback]) => {
      if (
        (conditions.callee && !conditions.callee(node.callee)) ||
        (conditions.arguments && !conditions.arguments(node.arguments))
      ) {
        return;
      }

      return callback(node, this);
    });

    getAffectedNodes(node, this).forEach((affectedNode) => {
      this.graph.addEdge(affectedNode, node);
      if (t.isIdentifier(affectedNode)) {
        this.graph.addEdge(
          this.scope.getDeclaration(affectedNode)!,
          affectedNode
        );
      }
    });
  },

  /*
   * SequenceExpression
   * It is a special case of expression in which the value of the whole
   * expression depends only on the last subexpression in the list.
   * The rest of the subexpressions can be omitted if they don't have dependent nodes.
   *
   * Example:
   * const a = (1, 2, b = 3, 4, b + 2); // `a` will be equal 5
   */
  SequenceExpression(this: GraphBuilderState, node: SequenceExpression) {
    // Sequence value depends on only last expression in the list
    this.baseVisit(node, true);
    if (node.expressions.length > 0) {
      this.graph.addEdge(node, node.expressions[node.expressions.length - 1]);
    }
  },
};

export const identifierHandlers: IdentifierHandlers = {
  declare: [
    ['CatchClause', 'param'],
    ['Function', 'params'],
    ['FunctionExpression', 'id'],
    ['RestElement', 'argument'],
    ['ThrowStatement', 'argument'],
    ['VariableDeclarator', 'id'],
  ],
  keep: [['ObjectProperty', 'key']],
  refer: [
    ['ArrayExpression', 'elements'],
    ['AssignmentExpression', 'left', 'right'],
    ['BinaryExpression', 'left', 'right'],
    ['CallExpression', 'arguments', 'callee'],
    ['ConditionalExpression', 'test', 'consequent', 'alternate'],
    ['ForInStatement', 'right'],
    ['Function', 'body'],
    ['IfStatement', 'test'],
    ['LogicalExpression', 'left', 'right'],
    ['NewExpression', 'arguments', 'callee'],
    ['ObjectProperty', 'value'],
    ['ReturnStatement', 'argument'],
    ['SequenceExpression', 'expressions'],
    ['SwitchStatement', 'discriminant'],
    ['UnaryExpression', 'argument'],
    ['UpdateExpression', 'argument'],
    ['VariableDeclarator', 'init'],
  ],
};
