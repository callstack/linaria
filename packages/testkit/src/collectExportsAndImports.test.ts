/* eslint-env jest */
import { join } from 'path';

import * as babel from '@babel/core';
import type { NodePath } from '@babel/traverse';
import traverse from '@babel/traverse';
import type { Program } from '@babel/types';
import dedent from 'dedent';
import * as ts from 'typescript';

import { collectExportsAndImports } from '@linaria/babel-preset';
import type { IImport, IState } from '@linaria/babel-preset';

function typescriptCommonJS(source: string): string {
  const result = ts.transpileModule(source, {
    compilerOptions: { module: ts.ModuleKind.CommonJS },
  });

  return result.outputText;
}

function babelCommonJS(source: string): string {
  const result = babel.transformSync(source, {
    babelrc: false,
    filename: join(__dirname, 'source.ts'),
    presets: [
      [
        require.resolve('@babel/preset-env'),
        {
          targets: 'ie 11',
        },
      ],
      '@babel/preset-typescript',
    ],
  });

  return result?.code ?? '';
}

function babelNode16(source: string): string {
  const result = babel.transformSync(source, {
    babelrc: false,
    filename: join(__dirname, 'source.ts'),
    presets: [
      [
        '@babel/preset-typescript',
        {
          onlyRemoveTypeImports: true,
        },
      ],
    ],
  });

  return result?.code ?? '';
}

const compilers: [name: string, compiler: (code: string) => string][] = [
  ['as is', babelNode16],
  ['babelCommonJS', babelCommonJS],
  ['typescriptCommonJS', typescriptCommonJS],
];

function run(compiler: (code: string) => string, code: string) {
  const compiled = compiler(code);
  const filename = join(__dirname, 'source.ts');
  expect(compiled).toMatchSnapshot('compiled with');

  const ast = babel.parse(compiled, {
    babelrc: false,
    filename,
  });

  let collected: IState | undefined;

  traverse(ast, {
    Program: {
      enter(path: NodePath<Program>) {
        collected = collectExportsAndImports(path, filename);
      },
    },
  });

  return (
    collected ?? {
      imports: [],
      exports: [],
    }
  );
}

const safeResolve = (name: string): string => {
  try {
    return require.resolve(name);
  } catch (e: unknown) {
    return name;
  }
};

const findBySource = (imports: IImport[], source: string) => {
  const resolved = safeResolve(source);
  return imports
    .filter((item) => item.source === resolved)
    .map((item) => item.imported)
    .sort();
};

describe.each(compilers)('collectExportsAndImports (%s)', (name, compiler) => {
  it('imports', () => {
    const { imports } = run(
      compiler,
      dedent`
        import unknownDefault, { unknown, another as unknownRenamed } from 'unknown-package';
        import type types from '@linaria/types';
        import atomic from '@linaria/atomic';
        import * as ns from '@linaria/namespace';
        import * as linaria from '@linaria/core';
        import {
          test,
          another as customName,
          type Styled
        } from '@linaria/react';

        const { only, few, fields } = linaria;

        const bar = linaria.bar;

        export { unknownDefault, unknown, unknownRenamed, types, atomic, ns, test, customName, Styled, only, few, fields, bar };
      `
    );

    expect(imports).toHaveLength(11);

    const find = (source: string) => findBySource(imports, source);

    expect(find('unknown-package')).toEqual(['=', 'another', 'unknown']);

    expect(find('@linaria/types')).toHaveLength(0);

    expect(find('@linaria/namespace')).toEqual(['*']);

    expect(find('@linaria/atomic')).toEqual(['=']);

    expect(find('@linaria/core')).toEqual(['bar', 'few', 'fields', 'only']);

    expect(find('@linaria/react')).toEqual(['another', 'test']);
  });

  it('requires', () => {
    const { imports } = run(
      compiler,
      dedent`
        const notModule = (() => {
          const require = () => ({});
          const { dep } = require('@linaria/shaker');
          return result;
        })();
        const module = (() => {
          const { dep } = require('@linaria/something');
          return result;
        })();
        const fullNamespace = require('@linaria/shaker');
        const { named } = require('@linaria/shaker');
        const { ...unknownRest } = require('@linaria/unknown');

        export { notModule, module, fullNamespace, named, unknownRest };
      `
    );

    const find = (source: string) => findBySource(imports, source);

    expect(imports).toHaveLength(4);

    expect(find('@linaria/something')).toEqual(['dep']);
    expect(find('@linaria/shaker')).toEqual(['*', 'named']);
    expect(find('@linaria/unknown')).toEqual(['*']);
  });

  xit('dynamic imports', () => {
    const { imports } = run(
      compiler,
      dedent`
        const fullNamespace = import('@linaria/shaker');
        const { named } = await import('@linaria/shaker');
        const { ...unknownRest } = await import('@linaria/unknown');

        export { fullNamespace, named, unknownRest };
      `
    );

    const find = (source: string) => findBySource(imports, source);

    expect(imports).toHaveLength(3);

    expect(find('@linaria/shaker')).toEqual(['*', 'named']);
    expect(find('@linaria/unknown')).toEqual(['*']);
  });

  xit('exports', () => {
    const { exports } = run(
      compiler,
      dedent`
        const a = 1;

        export { a };

        export const b = 2;

        export default function () {};
      `
    );

    expect(exports.map((i) => i.exported)).toHaveLength(4);

    // expect(find('@linaria/something')).toEqual(['dep']);
    // expect(find('@linaria/shaker')).toEqual(['*', 'named']);
    // expect(find('@linaria/unknown')).toEqual(['*']);
  });

  xit('re-exports', () => {
    const { exports } = run(
      compiler,
      dedent`
        export * from "module1";
        export * as name7 from "module2";
        export { name8, name9, name10 } from "module3";
        export { import1 as name11, import2 as name12, name13 } from "module4";
        export { default } from "module5";
      `
    );

    expect(exports).toHaveLength(4);

    // expect(find('@linaria/something')).toEqual(['dep']);
    // expect(find('@linaria/shaker')).toEqual(['*', 'named']);
    // expect(find('@linaria/unknown')).toEqual(['*']);
  });
});
