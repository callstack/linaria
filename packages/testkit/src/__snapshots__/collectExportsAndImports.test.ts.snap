// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collectExportsAndImports (as is) imports: compiled with 1`] = `
"import unknownDefault, { unknown, another as unknownRenamed } from 'unknown-package';
import atomic from '@linaria/atomic';
import * as ns from '@linaria/namespace';
import * as linaria from '@linaria/core';
import { test, another as customName } from '@linaria/react';
const {
  only,
  few,
  fields
} = linaria;
const bar = linaria.bar;
export { unknownDefault, unknown, unknownRenamed, atomic, ns, test, customName, only, few, fields, bar };"
`;

exports[`collectExportsAndImports (as is) requires: compiled with 1`] = `
"const notModule = (() => {
  const require = () => ({});

  const {
    dep
  } = require('@linaria/shaker');

  return result;
})();

const module = (() => {
  const {
    dep
  } = require('@linaria/something');

  return result;
})();

const fullNamespace = require('@linaria/shaker');

const {
  named
} = require('@linaria/shaker');

const { ...unknownRest
} = require('@linaria/unknown');

export { notModule, module, fullNamespace, named, unknownRest };"
`;

exports[`collectExportsAndImports (babelCommonJS) imports: compiled with 1`] = `
"\\"use strict\\";

function _typeof(obj) { \\"@babel/helpers - typeof\\"; return _typeof = \\"function\\" == typeof Symbol && \\"symbol\\" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && \\"function\\" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? \\"symbol\\" : typeof obj; }, _typeof(obj); }

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
Object.defineProperty(exports, \\"atomic\\", {
  enumerable: true,
  get: function get() {
    return _atomic.default;
  }
});
exports.bar = void 0;
Object.defineProperty(exports, \\"customName\\", {
  enumerable: true,
  get: function get() {
    return _react.another;
  }
});
exports.only = exports.ns = exports.fields = exports.few = void 0;
Object.defineProperty(exports, \\"test\\", {
  enumerable: true,
  get: function get() {
    return _react.test;
  }
});
Object.defineProperty(exports, \\"unknown\\", {
  enumerable: true,
  get: function get() {
    return _unknownPackage.unknown;
  }
});
Object.defineProperty(exports, \\"unknownDefault\\", {
  enumerable: true,
  get: function get() {
    return _unknownPackage.default;
  }
});
Object.defineProperty(exports, \\"unknownRenamed\\", {
  enumerable: true,
  get: function get() {
    return _unknownPackage.another;
  }
});

var _unknownPackage = _interopRequireWildcard(require(\\"unknown-package\\"));

var _atomic = _interopRequireDefault(require(\\"@linaria/atomic\\"));

var ns = _interopRequireWildcard(require(\\"@linaria/namespace\\"));

exports.ns = ns;

var linaria = _interopRequireWildcard(require(\\"@linaria/core\\"));

var _react = require(\\"@linaria/react\\");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== \\"function\\") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== \\"object\\" && typeof obj !== \\"function\\") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== \\"default\\" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var only = linaria.only,
    few = linaria.few,
    fields = linaria.fields;
exports.fields = fields;
exports.few = few;
exports.only = only;
var bar = linaria.bar;
exports.bar = bar;"
`;

exports[`collectExportsAndImports (babelCommonJS) requires: compiled with 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.unknownRest = exports.notModule = exports.named = exports.module = exports.fullNamespace = void 0;

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var notModule = function () {
  var require = function require() {
    return {};
  };

  var _require = require('@linaria/shaker'),
      dep = _require.dep;

  return result;
}();

exports.notModule = notModule;

var _module = function () {
  var _require2 = require('@linaria/something'),
      dep = _require2.dep;

  return result;
}();

exports.module = _module;

var fullNamespace = require('@linaria/shaker');

exports.fullNamespace = fullNamespace;

var _require3 = require('@linaria/shaker'),
    named = _require3.named;

exports.named = named;

var _require4 = require('@linaria/unknown'),
    unknownRest = _extends({}, _require4);

exports.unknownRest = unknownRest;"
`;

exports[`collectExportsAndImports (typescriptCommonJS) imports: compiled with 1`] = `
"\\"use strict\\";
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.bar = exports.fields = exports.few = exports.only = exports.customName = exports.test = exports.ns = exports.atomic = exports.unknownRenamed = exports.unknown = exports.unknownDefault = void 0;
var unknown_package_1 = require(\\"unknown-package\\");
exports.unknownDefault = unknown_package_1.default;
Object.defineProperty(exports, \\"unknown\\", { enumerable: true, get: function () { return unknown_package_1.unknown; } });
Object.defineProperty(exports, \\"unknownRenamed\\", { enumerable: true, get: function () { return unknown_package_1.another; } });
var atomic_1 = require(\\"@linaria/atomic\\");
exports.atomic = atomic_1.default;
var ns = require(\\"@linaria/namespace\\");
exports.ns = ns;
var linaria = require(\\"@linaria/core\\");
var react_1 = require(\\"@linaria/react\\");
Object.defineProperty(exports, \\"test\\", { enumerable: true, get: function () { return react_1.test; } });
Object.defineProperty(exports, \\"customName\\", { enumerable: true, get: function () { return react_1.another; } });
var only = linaria.only, few = linaria.few, fields = linaria.fields;
exports.only = only;
exports.few = few;
exports.fields = fields;
var bar = linaria.bar;
exports.bar = bar;
"
`;

exports[`collectExportsAndImports (typescriptCommonJS) requires: compiled with 1`] = `
"\\"use strict\\";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === \\"function\\")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, \\"__esModule\\", { value: true });
exports.unknownRest = exports.named = exports.fullNamespace = exports.module = exports.notModule = void 0;
var notModule = (function () {
    var require = function () { return ({}); };
    var dep = require('@linaria/shaker').dep;
    return result;
})();
exports.notModule = notModule;
var module = (function () {
    var dep = require('@linaria/something').dep;
    return result;
})();
exports.module = module;
var fullNamespace = require('@linaria/shaker');
exports.fullNamespace = fullNamespace;
var named = require('@linaria/shaker').named;
exports.named = named;
var unknownRest = __rest(require('@linaria/unknown'), []);
exports.unknownRest = unknownRest;
"
`;
